ssg/rule_dir_stats.py

```diff
@@ -0,0 +1,462 @@
+"""
+This module contains common code shared by utils/rule_dir_stats.py and
+utils/rule_dir_diff.py. This code includes functions for walking the output
+of the utils/rule_dir_json.py script, and filtering functions used in both
+scripts.
+"""
+
+from __future__ import absolute_import
+from __future__ import print_function
+
+import os
+from collections import defaultdict
+
+from .build_remediations import REMEDIATION_TO_EXT_MAP as REMEDIATION_MAP
+from .utils import subset_dict
+
+
+def get_affected_products(rule_obj):
+    """
+    From a rule_obj, return the set of affected products from rule.yml
+    """
+    return set(rule_obj['products'])
+
+
+def get_all_affected_products(args, rule_obj):
+    """
+    From a rule_obj, return the set of affected products from rule.yml, and
+    all fixes and checks.
+
+    If args.strict is set, this function is equivalent to
+    get_affected_products. Otherwise, it includes ovals and fix content based
+    on the values of args.fixes_only and args.ovals_only.
+    """
+
+    affected_products = get_affected_products(rule_obj)
+
+    if args.strict:
+        return affected_products
+
+    if not args.fixes_only:
+        for product in rule_obj['oval_products']:
+            affected_products.add(product)
+
+    if not args.ovals_only:
+        for product in rule_obj['remediation_products']:
+            affected_products.add(product)
+
+    return affected_products
+
+
+def _walk_rule(args, rule_obj, oval_func, remediation_func, verbose_output):
+    """
+    Walks a single rule and updates verbose_output if visited. Returns visited
+    state as a boolean.
+
+    Internal function for walk_rules and walk_rules_parallel.
+    """
+
+    rule_id = rule_obj['id']
+
+    affected_products = get_all_affected_products(args, rule_obj)
+    if not affected_products.intersection(args.products):
+        return False
+    if args.query and rule_id not in args.query:
```

Comment:
What do you think of also considering substrings of a rule id?
For example, I would like to query all audit rules related to kernel modules, so `-q audit_rules_kernel_module`.